src/Core/Localization/Currency/DataLayer/CurrencyDatabase.php

```diff
@@ -0,0 +1,144 @@
+    }
+
+    /**
+     * Set the lower layer.
+     * When reading data, if nothing is found then it will try to read in the lower data layer
+     * When writing data, the data will also be written in the lower data layer.
+     *
+     * @param currencyDataLayerInterface $lowerLayer
+     *                                               The lower data layer
+     *
+     * @return self
+     */
+    public function setLowerLayer(CurrencyDataLayerInterface $lowerLayer)
+    {
+        $this->lowerDataLayer = $lowerLayer;
+
+        return $this;
+    }
+
+    /**
+     * Actually read a data object into the current layer.
+     *
+     * Data is read into database
+     *
+     * @param LocalizedCurrencyId $currencyDataId
+     *                                            The CurrencyData object identifier (currency code + locale code)
+     *
+     * @return CurrencyData|null
+     *                           The wanted CurrencyData object (null if not found)
+     *
+     * @throws LocalizationException
+     *                               When $currencyDataId is invalid
+     */
+    protected function doRead($currencyDataId)
+    {
+        if (!$currencyDataId instanceof LocalizedCurrencyId) {
+            throw new LocalizationException('$currencyDataId must be a CurrencyDataIdentifier object');
+        }
+
+        $localeCode = $currencyDataId->getLocaleCode();
+        $currencyCode = $currencyDataId->getCurrencyCode();
+        $currencyEntity = $this->dataProvider->getCurrencyByIsoCode($currencyCode, $localeCode);
+
+        if (null === $currencyEntity) {
+            return null;
+        }
+
+        $currencyData = new CurrencyData();
+
+        $currencyData->setIsoCode($currencyEntity->iso_code);
+        $currencyData->setNames([$localeCode => $currencyEntity->name]);
+        $currencyData->setNumericIsoCode($currencyEntity->numeric_iso_code);
+        $currencyData->setSymbols([$localeCode => $currencyEntity->symbol]);
+        $currencyData->setPrecision($currencyEntity->precision);
+
+        return $currencyData;
+    }
+
+    /**
+     * Actually write a data object into the current layer
+     * Here, this is a DB insert/update...
+     *
+     * @param LocalizedCurrencyId $currencyDataId
+     *                                            The CurrencyData object identifier (currency code + locale code)
+     * @param CurrencyData $currencyData
+     *                                   The data object to be written
+     *
+     * @throws DataLayerException
+     *                            If something goes wrong when trying to write into DB
+     * @throws LocalizationException
+     *                               When $currencyDataId is invalid
+     */
+    protected function doWrite($currencyDataId, $currencyData)
+    {
+        if (!$currencyDataId instanceof LocalizedCurrencyId) {
+            throw new LocalizationException('$currencyDataId must be a CurrencyDataIdentifier object');
```

Comment:
Prefer `'First parameter must be and instance of ' . LocalizedCurrencyId::class`